cmd/web-repo-proxy/main.go

```diff
@@ -0,0 +1,361 @@
+
+	// Write out the given data to the file system.
+	for filename, content := range data {
+		filePath := filepath.Join(repositoryRoot, filename)
+		err = ioutil.WriteFile(filePath, []byte(content), 0644)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Set up the git repository.
+	gitCommands := [][]string{
+		[]string{"init"},
+		[]string{"add", "."},
+		[]string{"commit", "-m", "Initial commit"},
+		// update-server-info is needed to serve repositories over static http.
+		[]string{"update-server-info"},
+	}
+	for _, command := range gitCommands {
+		cmd := exec.Command("git", command...)
+		cmd.Dir = repositoryRoot
+		err = cmd.Run()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Choose the real repository path and move it there.
+	for {
+		repository := repository{
+			key:  createRandomKey(),
+			name: name,
+		}
+		newRoot := repository.filePathRoot()
+		err = os.Rename(repositoryRoot, newRoot)
+		if err != nil {
+			return nil, err
+		}
+		repositoryRoot = newRoot
+		// Corner case: only finish if we picked a unique key.
+		repos := repositoriesForKey(repository.key)
+		if len(repos) <= 1 {
+			return &repository, nil
+		}
+	}
+}
+
+// Check whether any repositories have expired, and delete them if so.
+func deleteExpiredRepositories() {
+	curTime := time.Now()
+	files, err := ioutil.ReadDir(repositoriesRoot)
+	if err != nil {
+		log15.Error("Couldn't read repositories directory", "error", err)
+		return
+	}
+	for _, file := range files {
+		if file.ModTime().Add(repositoryTTL).Before(curTime) {
+			// This repository has expired.
+			tempDir, err := ioutil.TempDir("", "web-repo-proxy")
+			if err != nil {
+				log15.Error("Couldn't create temporary directory", "error", err)
+				continue
+			}
+			fromPath := filepath.Join(repositoriesRoot, file.Name())
+			toPath := filepath.Join(tempDir, file.Name())
+			err = os.Rename(fromPath, toPath)
+			if err != nil {
+				log15.Error("Couldn't move expired repository entry", "error", err)
+				continue
+			}
+			err = os.RemoveAll(tempDir)
+			if err != nil {
+				log15.Error("Couldn't delete temporary directory", "tempDir", tempDir)
+			}
+		}
+	}
+}
+
+// Watches the repository deletion queue and removes repositories once
+// their goodUntil field is past.
+func repositoryDeleter() {
+	for {
+		deleteExpiredRepositories()
+		time.Sleep(30 * time.Second)
+	}
+}
+
+type CreateRepositoryRequest struct {
+	RepositoryName string            `json:"repositoryName"`
+	FileContents   map[string]string `json:"fileContents"`
+}
+
+type CreateRepositoryResponse struct {
+	UrlPath   string `json:"urlPath"`
+	GoodUntil int64  `json:"goodUntil"`
+}
+
+// Given a Reader with a create-repository request, return the corresponding
+// CreateRepositoryRequest with guaranteed-safe values, or an error.
+func sanitizedCreateRequest(requestBody io.Reader) (*CreateRepositoryRequest, error) {
+	decoder := json.NewDecoder(requestBody)
+	var request CreateRepositoryRequest
+	err := decoder.Decode(&request)
```

Comment:
no need for the intermediate decoder variable

```suggestion
	err := json.NewDecoder(requestBody).Decode(&request)
```